Add a suffix to the material output auxvariables if the same name is already in use

GiudGiud opened the pull request:
Closes #17905

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/766656/) on 39dc450 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/18182/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/18182/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 36ccaabf58e66ce75a1723d604deeb1b360a34f1`

GiudGiud committed:
Add exodus output of vector material properties
Allow warnings in truss test that relies on material property output for its calculation, refs #18371

loganharbour commented:
I would prefer that variable names not change under the hood.

How about instead, if a duplicate is found throw an error. Add a `prefix` param to the output object and in the error say that you must set the prefix.

GiudGiud commented:
This is a little more code for a very simple thing. 
I ll let you look at the latest commit.

I like a suffix more for this issue since outputs are ordered alphabetically in paraview.

For the sake of argument,
This is just for output, you're not supposed to be using these aux-variables in your simulation. Not to mention that there is a perfectly viable variable and a material property of the same name you can be using.

Also added output of AD vector properties. 
Please let me know when the design is fine for you, I ll add tests then

loganharbour commented:
> For the sake of argument,
This is just for output, you're not supposed to be using these aux-variables in your simulation. Not to mention that there is a perfectly viable variable and a material property of the same name you can be using.

Sorry, I'm not meaning to be difficult here. I would just prefer that these changes don't happen under the hood without any kind of clear notification. Simple case: someone has a test that tests material properties, so it uses the output properties debug. Then, they go add something with the same variable and all of a sudden their solve still works but just diffs

loganharbour commented on framework/src/materials/MaterialBase.C:
```diff
@@ -52,6 +52,8 @@ MaterialBase::validParams()
       "output_properties",
       "List of material properties, from this material, to output (outputs "
       "must also be defined to an output type)");
+  params.addParam<std::string>("material_output_suffix", "A suffix to disambiguate variables and "
```
I would improve on this description just a bit

loganharbour reviewed (commented):
Looks good to me

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/766769/) on cf9c26a : invalidated by @loganharbour

Fixed precheck

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/766769/) on cf9c26a wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/18182/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/18182/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 36ccaabf58e66ce75a1723d604deeb1b360a34f1`

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/766872/) on 184a375 wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/18182/site)

This comment will be updated on new commits.

GiudGiud commented:
Had to retold some tests that were using existing aux variables to output material properties. 
If there is an option to change the names of compared variables in ExoDiff we could avoid that. After a quick look at the manual didn't seem like there is

bwspenc reviewed (dismissed):
Looks good to me.  I agree with the behavior of giving an error if there's a conflict rather than renaming it behind the scenes.

dschwen commented on framework/src/actions/MaterialOutputAction.C:
```diff
@@ -262,14 +282,19 @@ MaterialOutputAction::getParams(const std::string & type,
                                 const MaterialBase & material)
 {
+  // Use a suffix to disambiguate variables and material properties with the same name
+  const auto output_suffix = material.isParamValid("material_output_suffix")
```
Any reason you don't just default this param to `""`?

GiudGiud replied:
not really. I changed it to be default '' now

dschwen commented on framework/src/actions/MaterialOutputAction.C:
```diff
@@ -340,6 +365,31 @@ MaterialOutputAction::materialOutputHelper<RealVectorValue>(const std::string &
+  std::vector<std::string> names(3);
+  for (unsigned int i = 0; i < LIBMESH_DIM; ++i)
+  {
+    std::ostringstream oss;
+    oss << property_name << "_" << suffix[i];
+    names[i] = oss.str();
```
I think this should work here, too
```suggestion
    names[i] = property_name + '_' + suffix[i];
```

dschwen reviewed (commented):
Use of stringstream doesn't seem necessary, use of parameter default would simplify code a bit.

GiudGiud commented:
Closing as the need is covered by declare_suffix introduced by CNSFV

GiudGiud commented:
Actually we should still have at least a warning. The existence of the suffix parameter isnt enough

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/778310/) on f8299d5 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/18182/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/18182/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format a969d327c03b526dd25c1af5b93f3b3aeabe2814`

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/778337/) on cda84d8 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/18182/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/18182/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format a969d327c03b526dd25c1af5b93f3b3aeabe2814`

GiudGiud commented:
and exodus output of vector matprops is worth it too

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/778364/) on ad2ad9f wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/18182/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/18182/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format a969d327c03b526dd25c1af5b93f3b3aeabe2814`

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/778364/) on ad2ad9f wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/18182/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/18182/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format a969d327c03b526dd25c1af5b93f3b3aeabe2814`

GiudGiud commented:
Regold could be avoided at the cost of an allow_warning on these tests. 
Any preference?

GiudGiud commented:
Ok I removed the re-gold because this isnt enough. 
See #18371 for the full details. 
Imo this is good to go

moosebuild commented:
Job [End User Experience (CentOS 8)](https://civet.inl.gov/job/779096/) on 536e25e : invalidated by @GiudGiud

Only pgh test left, there&#39;&#39;s a PR there to fix it

loganharbour commented:
In these cases we try to fix it in the offending app temporarily so that we don't have to break things in the process (two PRs needed in the target app)

GiudGiud commented:
I dont think it's possible here. 
We could break this PR in two though, that could pass the tests.

moosebuild commented:
Job [Internal App tests](https://civet.inl.gov/job/779087/) on 536e25e : invalidated by @GiudGiud

pgh is fixed now

moosebuild commented:
Job [Internal App tests](https://civet.inl.gov/job/779087/) on 536e25e : invalidated by @GiudGiud

pgh fixed for real now?

GiudGiud commented:
ready for re-review or merging

GiudGiud commented:
@loganharbour @dschwen @bwspenc need a checkmark to merge this pls

bwspenc commented on framework/src/actions/MaterialOutputAction.C:
```diff
@@ -200,6 +204,11 @@ MaterialOutputAction::act()
     {
       oss << "\n  " << var_name;
+      // Warn if adding material properties auxiliary variables with the same name as a variable
+      if (_problem->hasVariable(var_name))
```
I haven't really followed this closely, but I'm having trouble understanding why this is a warning rather than an error. It looks like if it already exists, we create it anyway (which I guess works), and then have two things potentially setting its value. Why not just make that an error?

GiudGiud replied:
because the truss tests rely on it. When someone fixes them, I raised an issue, we can convert to an error. 
A warning is better than nothing, the current behavior.

loganharbour replied:
Why can we not fix those tests?

loganharbour replied:
I too would prefer to get the error in on the initial PR instead of having to do it in the future

GiudGiud replied:
I've looked into changing and if we make it an error then we wont be able to recover simulations that have a material output. The first run will create the variables, the second one will error out on already existing variables. We could have people add a prefix in between, but it's not worth erroring out their simulation tbh

GiudGiud replied:
looks like the PR is borked anyway. I ll get around to fixing it

GiudGiud commented:
Force pushed to improve the commit messages, but this is still ready.

GiudGiud committed:
Add warning if material output uses the same variable name as the material name, refs #17905

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/797985/) on a21ca8d wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/18182/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/18182/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format ed0fae0285a385b486af09f484c13f817c04e708`

GiudGiud commented:
ok should be ready now.

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/798025/) on cd51a04 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/18182/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/18182/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format ed0fae0285a385b486af09f484c13f817c04e708`

GiudGiud commented:
I'd be down for this to merged before more truss stuff with likely the same issue is added in #18269
@loganharbour

GiudGiud commented:
umm lemme look at those

GiudGiud commented:
ah in the 30 days that this has been sitting there, people have added more problematic tests

GiudGiud committed:
Address review:
- make it an error to add a variable with the same name as an existing one with the material output action
- fixup truss tests to make them avoid that scenario
refs #17905
closes #18371

Add two tests for using the AD vector material property output

GiudGiud commented:
Changed from a warning to an error. 
My analysis of truss tests was wrong, they could be fixed easily, just needed to play with the output parameters.
@loganharbour should be good now

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/919886/) on f37dc00 wanted to post the following:


## Framework coverage

<table>
  <thead>
    <tr>
      <th></th>
      <th>ed4553</th>
      <th colspan=4>#18182 f37dc0</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>82.18%</td>
      <td>82.16%</td>
      <td>-0.01%</td>
      <td>18.75%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>72852</td>
      <td>72854</td>
      <td>+2</td>
      <td>3</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>15802</td>
      <td>15816</td>
      <td>+14</td>
      <td>13</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/18182/coverage_diff/framework.html)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/18182/coverage/framework)


## Modules coverage

Coverage did not change

## Full coverage reports

<details>
<summary>Reports</summary>

- [`framework`](https://mooseframework.inl.gov/docs/PRs/18182/coverage/framework)
- [`chemical_reactions`](https://mooseframework.inl.gov/docs/PRs/18182/coverage/chemical_reactions)
- [`combined`](https://mooseframework.inl.gov/docs/PRs/18182/coverage/combined)
- [`contact`](https://mooseframework.inl.gov/docs/PRs/18182/coverage/contact)
- [`external_petsc_solver`](https://mooseframework.inl.gov/docs/PRs/18182/coverage/external_petsc_solver)
- [`fluid_properties`](https://mooseframework.inl.gov/docs/PRs/18182/coverage/fluid_properties)
- [`fsi`](https://mooseframework.inl.gov/docs/PRs/18182/coverage/fsi)
- [`functional_expansion_tools`](https://mooseframework.inl.gov/docs/PRs/18182/coverage/functional_expansion_tools)
- [`geochemistry`](https://mooseframework.inl.gov/docs/PRs/18182/coverage/geochemistry)
- [`heat_conduction`](https://mooseframework.inl.gov/docs/PRs/18182/coverage/heat_conduction)
- [`level_set`](https://mooseframework.inl.gov/docs/PRs/18182/coverage/level_set)
- [`misc`](https://mooseframework.inl.gov/docs/PRs/18182/coverage/misc)
- [`navier_stokes`](https://mooseframework.inl.gov/docs/PRs/18182/coverage/navier_stokes)
- [`peridynamics`](https://mooseframework.inl.gov/docs/PRs/18182/coverage/peridynamics)
- [`phase_field`](https://mooseframework.inl.gov/docs/PRs/18182/coverage/phase_field)
- [`porous_flow`](https://mooseframework.inl.gov/docs/PRs/18182/coverage/porous_flow)
- [`ray_tracing`](https://mooseframework.inl.gov/docs/PRs/18182/coverage/ray_tracing)
- [`rdg`](https://mooseframework.inl.gov/docs/PRs/18182/coverage/rdg)
- [`reactor`](https://mooseframework.inl.gov/docs/PRs/18182/coverage/reactor)
- [`richards`](https://mooseframework.inl.gov/docs/PRs/18182/coverage/richards)
- [`stochastic_tools`](https://mooseframework.inl.gov/docs/PRs/18182/coverage/stochastic_tools)
- [`tensor_mechanics`](https://mooseframework.inl.gov/docs/PRs/18182/coverage/tensor_mechanics)
- [`xfem`](https://mooseframework.inl.gov/docs/PRs/18182/coverage/xfem)

</details>

## Warnings

- `framework` new line coverage rate 18.75% is less than the suggested 90.0%



This comment will be updated on new commits.

GiudGiud commented:
Looks like this needs a test and an app patch

GiudGiud commented:
Patches to all 3 apps are in PRs, I'll ping back when they're in

moosebuild commented:
Job [Internal App tests](https://civet.inl.gov/job/920068/) on 5d0f709 : invalidated by @GiudGiud

all patches in

moosebuild commented:
Job [Internal App tests](https://civet.inl.gov/job/920068/) on 5d0f709 : invalidated by @GiudGiud

Pgh should propagate to main again
